crates/kitsune_p2p/kitsune_p2p/tests/performance.rs

```diff
@@ -0,0 +1,191 @@
+            .await
+            .unwrap();
+    }
+
+    // Wait for the nodes to discover each other before publishing
+    wait_for_connected(sender_a.clone(), agent_b.clone(), all_spaces[0].clone()).await;
+    wait_for_connected(sender_b.clone(), agent_a.clone(), all_spaces[0].clone()).await;
+
+    let basis = Arc::new(KitsuneBasis::new(vec![0; 32]));
+
+    // Create some test data, roughly partitioned across the spaces as 'no data', 'data from node a', 'data from node b'
+    for i in 0..100 {
+        match i % 3 {
+            0 => (), // Skip, don't create data in every space
+            1 => {
+                let use_space = &all_spaces[i % 10];
+                let test_data = TestHostOp::new(use_space.clone().into());
+                harness_a.op_store().write().push(test_data.clone());
+
+                sender_a
+                    .broadcast(
+                        use_space.clone(),
+                        basis.clone(),
+                        KitsuneTimeout::from_millis(5_000),
+                        BroadcastData::Publish {
+                            source: agent_a.clone(),
+                            op_hash_list: vec![test_data.into()],
+                            context: FetchContext::default(),
+                        },
+                    )
+                    .await
+                    .unwrap();
+            }
+            2 => {
+                let use_space = &all_spaces[i % 10];
+                let test_data = TestHostOp::new(use_space.clone().into());
+                harness_b.op_store().write().push(test_data.clone());
+
+                sender_b
+                    .broadcast(
+                        use_space.clone(),
+                        basis.clone(),
+                        KitsuneTimeout::from_millis(5_000),
+                        BroadcastData::Publish {
+                            source: agent_b.clone(),
+                            op_hash_list: vec![test_data.into()],
+                            context: FetchContext::default(),
+                        },
+                    )
+                    .await
+                    .unwrap();
+            }
+            _ => {
+                unreachable!("because maths");
+            }
+        }
+    }
+
+    // Wait for 30s to allow gossip to happen so we can measure the number of host calls
+    tokio::time::sleep(std::time::Duration::from_secs(30)).await;
+
+    let drained_events = harness_a.drain_legacy_host_events().await;
+
+    let put_agent_info_signed_count = drained_events
+        .iter()
+        .filter(|e| matches!(e, RecordedKitsuneP2pEvent::PutAgentInfoSigned { .. }))
+        .count();
+
+    put_agent_info_signed_count.assert_close_to(90, 5);
+
+    let query_agents_count = drained_events
+        .iter()
+        .filter(|e| matches!(e, RecordedKitsuneP2pEvent::QueryAgents { .. }))
+        .count();
+
+    query_agents_count.assert_close_to(13732, 500);
```

Comment:
They weren't on 0.2.2, so from 0.2.5 these calls should be less of a problem because there will be more concurrency allowed. Even then though, they are going through the `ghost_actor` which gets overwhelmed when other load starts to get applied to it like data arriving. So I think this is using up a significant amount of the available capacity for that too. I am working on reducing the number of calls at the moment